### PR TITLE
Firefox: Text selection, Carets, & Scrollbars

### PR DESCRIPTION
**Pull request by ZimCodes**
- Applies Doki definition themes for **carets**, **scrollbars**, & **text selection**. 
- Update styles for popup menu. 
- Fixes switches in popup menu not disabling properly for both secondary and dark.
- and more...

#### Description

This pull request attempts to solve issue #59.

> In Chrome the Scroll bar and marked Text changes with the Theme you chose. And my question is if its possible to have that feature in Firefox and if yes how to get it.

##### Scrollbar, Caret, Text Selection
Using `scrollbar-width` and `scrollbar-color`, the scrollbars can be modified without generating a more complex solution. These CSS properties are **only compatible for Firefox**. Of course there's a `::-webkit-scrollbar` variant which is compatible for Chromium browsers. The carets and text selection components also have their own CSS properties and selectors. For carets, `caret-color`. For text selection, `::selection`.

Instead of creating a `<style>` & placing it in the `<head>`, the previously mentioned CSS properties are added to the _.css_ 
& then are modified using CSS variables.

[Content scripts](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts) are used to load scripts onto [<all_urls>](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Match_patterns#%3Call_urls%3E) pages. In this case, the scripts injects the scrollbar, caret & text selection CSS properties onto all websites. The scrollbar and text selection *(caret is coupled with text selection)* are **enabled by default** and can be disabled in the *options page*.  

##### Popup Menu Styles
Some styling was added to the popup menu. Deactivated, (*not disabled & not checked*), switches receive a nice glow providing depth and recognition that this particular switch is toggleable. The glow depends on whether the theme is light or dark. The Doki logo also has a glowing effect when hovered. The same glow rules apply here. Additionally, the Doki logo transitions to a reddish color when hovered providing a visual affordance.

##### Popup Menu Functionality
In accordance to the glowing effect, clicking on the Doki logo navigates you directly to the *options page*. Makes for a smoother user experience. The switches are now properly disabled when certain background variants of a theme are not available. For instance, a theme might have only a dark variant, therefore the secondary and dark switches will be disabled as there are no other variant available for this theme. 

##### Honorable Mentions
- *Random mode* was not functioning so its been repaired.
- Removes deprecated [colors.toolbar_field_separator](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/theme) theme key.
- Adjusted _CONTRIBUTING.md_ a bit.

#### Motivation and Context

This provides the appropriate enhancements mentioned in issue #59.
#### Screenshots (if appropriate):
Screenshot (14)
Screenshot (15)
Screenshot

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Non-Functional Change (non-user facing changes)

#### Checklist:

- [x] I updated the version (base package.json and re-built the master extension).
- [x] I updated the changelog with the new functionality.

### CODE REVIEW

**Comment by Unthrottled** [comment]
Hey, Thank You! I'll probably be able to check this out later today or tomorrow.

File: firefoxThemes/popup/popup.js
Comment (Unthrottled) [inline firefoxThemes/popup/popup.js:108]: So this is a bit tricky.

The name used in the drop down is `displayName` however, if there are 2 themes with the same `displayName` then the `name` of the theme is used. There are:
- Hatsune Miku & Nakano Miku
- Tohsaka Rin & Shima Rin

Light & Dark themes of the same character have a name of {{Character Name}} Dark/Light, so to get those the display name is the same.

So to make this more correct, filter by display name && group.

Cause otherwise 2 different Mikus & Rins are going to be hanging out.

File: firefoxThemes/waifus/Aqua/theme.json
Comment (Unthrottled) [inline firefoxThemes/waifus/Aqua/theme.json:138]: Probably a good idea to remove this from the `firefox.theme.template.json`, because next time I run a `buildThemes` it is going to put it right back.

File: CONTRIBUTING.md
Comment (Unthrottled) [inline CONTRIBUTING.md:205]: Thanks for proof-reading!

I've copy pasted this from other places, which means that this in incorrect in a bunch of other places, yayyy....... :sweat:

**Review by Unthrottled — CHANGES_REQUESTED** [review]
Very nice! I missed this functionality, thank you for spending the time to work on this.

File: firefoxThemes/popup/popup.js
Comment (Unthrottled) [reply firefoxThemes/popup/popup.js:24]: FYI, I actually did the icon theming in #64 . Will wait until this is in before I merge the other pull request in.

File: firefoxThemes/options/option.css
Comment (ZimCodes) [reply firefoxThemes/options/option.css:8]: Can you clarify this point? Should the background of the option page stay the default white color? or should it change based on the theme?

File: firefoxThemes/waifus/Aqua/theme.json
Comment (ZimCodes) [reply firefoxThemes/waifus/Aqua/theme.json:138]: As I am still a bit unfamiliar with the structure of the asset building, there is a *Kanna* light/dark background which is the same how does one remove it from the build tool?

File: firefoxThemes/options/option.css
Comment (ZimCodes) [reply firefoxThemes/options/option.css:8]: default
At the moment, the options page is consistently this color combo as shown above regardless of theme choice.

File: firefoxThemes/popup/popup.js
Comment (ZimCodes) [reply firefoxThemes/popup/popup.js:24]: The design looks nice! [Image reference.](https://user-images.githubusercontent.com/15972415/132136186-96a16e2b-2d9e-4219-a578-1dcee320d27d.png) Will the leftmost heart be used as an affordance for something?

File: firefoxThemes/options/option.css
Comment (Unthrottled) [reply firefoxThemes/options/option.css:2]: > should it change based on the theme?

Yes and I just realized the existing styling is inconsistent what I was wanting to change. Sorry about that.

image


> At the moment, the options page is consistently this color combo as shown above regardless of theme choice.

Yeah I noticed that too when I was testing, but when the tab is closed an then you go back to `manage extension` it becomes themed, right?

File: firefoxThemes/popup/popup.js
Comment (Unthrottled) [reply firefoxThemes/popup/popup.js:24]: > Image reference. Will the leftmost heart be used as an affordance for something?

I thought about it, but since there is already a `home` link provided , I did not see the  point of that taking the user home too. Unless people are used to clicking on icons to go home. (was assuming you were speaking to the image reference leftmost heart)

File: firefoxThemes/waifus/Aqua/theme.json
Comment (Unthrottled) [reply firefoxThemes/waifus/Aqua/theme.json:138]: > there is a Kanna primary/secondary background which is the same how does one remove it from the build tool?

That one is on me. The background get copied over from my asset repository during the `buildThemes` process. I can modify and test things out, because I probably am not going to make a different secondary Kanna background.

File: firefoxThemes/options/option.css
Comment (ZimCodes) [reply firefoxThemes/options/option.css:2]: Not in the version i have currently.

File: firefoxThemes/popup/popup.js
Comment (ZimCodes) [reply firefoxThemes/popup/popup.js:24]: My apologizes. I was referencing the heart located here.

File: firefoxThemes/waifus/Aqua/theme.json
Comment (ZimCodes) [reply firefoxThemes/waifus/Aqua/theme.json:138]: I see.

File: firefoxThemes/options/option.css
Comment (ZimCodes) [inline firefoxThemes/options/option.css:53]: Change the checkmark to instead be a heart icon. Naturally it changes color based on theme.

File: firefoxThemes/options/options.js
Comment (ZimCodes) [inline firefoxThemes/options/options.js]: I decided to use the `classNameColor` color definition instead as the other color definitions  made it difficult to see the checkmark in the checkboxes.

**Review by ZimCodes — COMMENTED** [review]
These are the recent updates and changes to options page.

**Comment by ZimCodes** [comment]
@Unthrottled I think that's about it for the changes unless there's something else I'm missing? I know that in the next pull request the -

> Could we use the transparency accent color for the blur/dropshadow effects in the popup, please?

has been completed. As well as the - 

> That one is on me. The background get copied over from my asset repository during the buildThemes process. I can modify and test things out, because I probably am not going to make a different secondary Kanna background.

- will be taken care in the build tool.

File: firefoxThemes/popup/popup.js
Comment (Unthrottled) [reply firefoxThemes/popup/popup.js:24]: Ah, so that is the `Heart Fairy` that follows your mouse around on "The Doki Theme Home" website.

https://doki-theme.unthrottled.io/?themeId=5de1c901-2a12-411f-b75b-64e43aaf2c9d 

Wanted to reference that, because I themed the logo in that project too.

**Comment by Unthrottled** [comment]
Alright, I'll have some time to look at this again tomorrow, thank you for these contributions!

File: firefoxThemes/popup/popup.js
Comment (ZimCodes) [reply firefoxThemes/popup/popup.js:24]: oh, I see now! Interesting!

Comment (ZimCodes) [reply firefoxThemes/popup/popup.js:24]: I recommend adding a fairy after effect movement trail. Or include a group of fairies that follow.

If you are going to include the previously mentioned effects keep an eye out on performance.

Comment (Unthrottled) [reply firefoxThemes/popup/popup.js:24]: > I recommend adding a fairy after effect movement trail. Or include a group of fairies that follow.

:laughing: I hadn't thought of adding particle effects on moments. Thanks for the suggestion!

File: firefoxThemes/options/option.css
Comment (Unthrottled) [reply firefoxThemes/options/option.css:53]: Interesting! I like the idea,

Are the heart's off center on your machine?

image

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js]: I like the idea! However I am a bit partial to the checkbox esthetic from the jetbrains plugin. 

image

What do you think about this?

Peek 2021-09-08 05-42

I can push the changes up if we both agree.

Comment (ZimCodes) [reply firefoxThemes/options/options.js]: Hmm... I'm more partial to the colored-in checkbox idea myself as it adds more coloration to the options page. Though the jetbrain aesthetic is fine too. If using the jetbrain's aesthetic, to keep things synonymous with the plugin, I recommend staying with the &#10003; over the &#9829; as the checkmark.

File: firefoxThemes/options/option.css
Comment (ZimCodes) [reply firefoxThemes/options/option.css:52]: Created a possible fix for the out-of-bounds checkmark. Not sure if it will succeed as debugging with the responsive design mode doesn't help out much reproducing the issue.

Comment (Unthrottled) [reply firefoxThemes/options/option.css:52]: Just checked again, and it's still off center. Might just be a Linux thing. Again, not too concerned.

File: firefoxThemes/options/options.js
Comment (Unthrottled) [reply firefoxThemes/options/options.js]: I'll leave it as is. I might play around with the idea of adopting the :heart: checkbox in other platforms.

**Review by Unthrottled — APPROVED** [review]
Changes look good, thank you for the contributions! :tada:

**Pull request merged**
